File: wlxx-services/wlxx-auth/wlxx-auth-provider/src/main/java/com/xzg/wlxx/auth/config/security/auth/AuthenticationService.java
package com.xzg.wlxx.auth.config.security.auth;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.lang.Assert;
import com.xzg.wlxx.auth.config.security.config.JwtTokenUtils;
import com.xzg.wlxx.common.base.ApiResult;
import com.xzg.wlxx.system.client.entity.dto.UserDto;
import com.xzg.wlxx.system.client.entity.po.TokenPo;
import com.xzg.wlxx.system.client.feign.TokenProvider;
import com.xzg.wlxx.system.client.feign.UserProvider;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class AuthenticationService {
    private final TokenProvider tokenService;
    private final UserProvider userService;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenUtils jwtService;
    private final AuthenticationManager authenticationManager;
    private final UserDetailsService userDetailsService;

    public ApiResult<?> register(AuthenticationDto request) {
        Assert.notBlank(request.getUsername(), "参数不能为空！");
        Assert.notBlank(request.getPassword(), "参数不能为空");

        var user = new UserDto();
        user.setUsername(request.getUsername());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        return ApiResult.success(userService.register(user));
    }

    public ApiResult<?> authenticate(AuthenticationDto dto) {

        UserDetails userDetails = userDetailsService.loadUserByUsername(dto.getUsername());

        if (userDetails == null || !passwordEncoder.matches(dto.getPassword(), userDetails.getPassword())) {
            return ApiResult.failure("用户名或密码不正确！");
        }

        Authentication authenticate = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(dto.getUsername(), dto.getPassword())
        );
        // TODO 考虑不从数据库中查数据

        userDetails = (SystemUserDetails) authenticate.getPrincipal();
        var user = userService.findByUsername(userDetails.getUsername());
        // var user = userService.findByEmail(request.getEmail());
        // var user = userDetailsService.loadUserByUsername(request.getEmail());


        var jwtToken = jwtService.generateToken(userDetails);
        var refreshToken = jwtService.generateRefreshToken(userDetails);
        revokeUserToken(user.getId());
        saveUserToken(user.getId(), jwtToken);
        AuthenticationVo authenticationVo = AuthenticationVo.builder()
                .accessToken(jwtToken)
                .refreshToken(refreshToken)
                .build();
        return ApiResult.success(authenticationVo);

    }

    public ApiResult<?> refreshToken(HttpServletRequest request,
                                     HttpServletResponse response) throws IOException {
        final String authHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
        final String refreshToken;
        final String username;
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return ApiResult.failure("你的操作非法！");
        }
        refreshToken = authHeader.substring(7);
        username = jwtService.extractUsername(refreshToken);
        if (username != null) {
            // log.info("username: {}",username);
            // TODO 考虑不从数据库获取用户数据
            // var user = userService.findByEmail(username);
            var userDetails = (SystemUserDetails) userDetailsService.loadUserByUsername(username);
            var user = userService.findByUsername(userDetails.getUsername());
            if (jwtService.isTokenValid(refreshToken, userDetails)) {
                var accessToken = jwtService.generateToken(userDetails);
                log.info("accessToken: {}", accessToken);
                revokeUserToken(user.getId());
                saveUserToken(user.getId(), accessToken);
                var authResponse = AuthenticationResponse.builder()
                        .accessToken(accessToken)
                        .refreshToken(refreshToken)
                        .build();
                // new ObjectMapper().writeValue(response.getOutputStream(), authResponse);
                return ApiResult.success(authResponse);
            }
            return ApiResult.failure("令牌无效！");
        }
        return ApiResult.failure("获取用户名失败！");
    }


    /**
     * 保存用户的 token
     *
     * @param userId
     * @param jwtToken
     */
    private void saveUserToken(Long userId, String jwtToken) {
        var token = TokenPo.builder()
                .token(jwtToken)
                .expired(false)
                .revoked(false)
                .userId(userId)
                .build();
        token.setId(userId);
        tokenService.saveOrUpdate(token);
    }


    /**
     * 销毁用户的 token
     * - expired = true
     * - revoked = true
     *
     * @param userId
     */
    private void revokeUserToken(Long userId) {
        var validUserToken = tokenService.findToken(TokenPo.builder()
                .userId(userId)
                .expired(true)
                .revoked(true)
                .build()
        );
        if (BeanUtil.isEmpty(validUserToken))
            return;
        validUserToken.setExpired(true);
        tokenService.saveOrUpdate(validUserToken);
    }
}

File: wlxx-services/wlxx-auth/wlxx-auth-provider/src/main/kotlin/com/xzg/wlxx/auth/config/security/auth/AuthenticationService.kt
package com.xzg.wlxx.auth.config.security.auth

import cn.hutool.core.bean.BeanUtil
import cn.hutool.core.lang.Assert
import com.xzg.wlxx.auth.config.security.config.JwtTokenUtils
import com.xzg.wlxx.common.base.ApiResult
import com.xzg.wlxx.common.base.ApiResult.Companion.failure
import com.xzg.wlxx.common.base.ApiResult.Companion.success
import com.xzg.wlxx.common.log.WlxxLog.Companion.log
import com.xzg.wlxx.system.client.entity.dto.RegisterUserDto
import com.xzg.wlxx.system.client.entity.po.TokenPo
import com.xzg.wlxx.system.client.feign.TokenProvider
import com.xzg.wlxx.system.client.feign.UserProvider
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.springframework.http.HttpHeaders
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.io.IOException

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
class AuthenticationService {
    private val tokenService: TokenProvider? = null
    private val userService: UserProvider? = null
    private val passwordEncoder: PasswordEncoder? = null
    private val jwtService: JwtTokenUtils? = null
    private val authenticationManager: AuthenticationManager? = null
    private val userDetailsService: UserDetailsService? = null
    fun register(request: AuthenticationDto): ApiResult<*> {
        Assert.notBlank(request.username, "参数不能为空！")
        Assert.notBlank(request.password, "参数不能为空")
        val user = RegisterUserDto(username = request.username, password = passwordEncoder!!.encode(request.password))

        return success<Any>(userService!!.register(user))
    }

    fun authenticate(dto: AuthenticationDto): ApiResult<*> {
        var userDetails = userDetailsService!!.loadUserByUsername(dto.username)
        if (userDetails == null || !passwordEncoder!!.matches(dto.password, userDetails.password)) {
            return failure<Any>("用户名或密码不正确！")
        }
        val authenticate = authenticationManager!!.authenticate(
            UsernamePasswordAuthenticationToken(dto.username, dto.password)
        )
        // TODO 考虑不从数据库中查数据
        userDetails = authenticate.principal as SystemUserDetails
        val user = userService!!.findByUsername(userDetails.getUsername())
        // var user = userService.findByEmail(request.getEmail());
        // var user = userDetailsService.loadUserByUsername(request.getEmail());
        val jwtToken = jwtService!!.generateToken(userDetails)
        val refreshToken = jwtService.generateRefreshToken(userDetails)
        user?.id?.let { revokeUserToken(it) }
        user?.id?.let { saveUserToken(it, jwtToken) }
        val authenticationVo = AuthenticationVo(accessToken = jwtToken, refreshToken = refreshToken)

        return success<Any>(authenticationVo)
    }

    @Throws(IOException::class)
    fun refreshToken(
        request: HttpServletRequest,
        response: HttpServletResponse?
    ): ApiResult<*> {
        val authHeader = request.getHeader(HttpHeaders.AUTHORIZATION)
        val refreshToken: String
        val username: String?
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return failure<Any>("你的操作非法！")
        }
        refreshToken = authHeader.substring(7)
        username = jwtService!!.extractUsername(refreshToken)
        if (username != null) {
            // log.info("username: {}",username);
            // TODO 考虑不从数据库获取用户数据
            // var user = userService.findByEmail(username);
            val userDetails = userDetailsService!!.loadUserByUsername(username) as SystemUserDetails
            val user = userService!!.findByUsername(userDetails.username)
            if (jwtService.isTokenValid(refreshToken, userDetails)) {
                val accessToken = jwtService.generateToken(userDetails)
                log.info("accessToken: {}", accessToken)
                user?.id?.let { revokeUserToken(it) }
                user?.id?.let { saveUserToken(it, accessToken) }
                val authResponse = AuthenticationResponse(accessToken = accessToken, refreshToken = refreshToken)

                // new ObjectMapper().writeValue(response.getOutputStream(), authResponse);
                return success<Any>(authResponse)
            }
            return failure<Any>("令牌无效！")
        }
        return failure<Any>("获取用户名失败！")
    }

    /**
     * 保存用户的 token
     *
     * @param userId
     * @param jwtToken
     */
    private fun saveUserToken(userId: Long, jwtToken: String?) {
        val token = TokenPo(jwtToken, false, false, userId)
        token.id = userId
        tokenService!!.saveOrUpdate(token)
    }

    /**
     * 销毁用户的 token
     * - expired = true
     * - revoked = true
     *
     * @param userId
     */
    private fun revokeUserToken(userId: Long) {
        val validUserToken = tokenService!!.findToken(
            TokenPo(userId = userId, expired = true, revoked = true)
        )
        if (BeanUtil.isEmpty(validUserToken)) return
        validUserToken?.expired = true
        tokenService.saveOrUpdate(validUserToken)
    }
}
